### ISSUE TITLE
[BUG] ncycles_n_i produces incorrect results

### ISSUE THREAD

**Issue by EliasBcd** [JuliaGraphs/LightGraphs.jl#1401]
**Description of the bug**
The function `ncycles_n_i(n, i)` that computes the maximal number of cycles of length `i` in a directed graph of size `n` has two issues:
1. It yields incorrect results, because the formula employed is actually incorrect;
2. It does cannot take into account self-loops.

**How to reproduce**
For instance 
```julia
ncycles_n_i(2, 2)
```
yield a result of 2, whereas we would expect a result of 1, as there is only one cycle of length 2 in a directed graph of length 2.
Additionally, we would expect `ncycles_n_i(n, 1)` to produce the number of self-loops is self-loops are allowed in the directed graph, and 0 otherwise. This is not the results.
For instance
```julia
ncycles_n_i(2, 1)
```

**Expected behavior**
It yields 1, whereas it should either yield 2 or 0, depending on whether self-loops are taken into account or not.

**Version information**
Current master

**Additional context**
The correct formula should be `binomial(n, i) * factorial(i-1)`, and not `binomial(n, n-i+1) *factorial(n-i)`.
Funnily enough, this does not impact the function `max_simple_cycles`, because, assuming there are not self loops, the new formula to compute `max_simple_cycles` is:
```julia
i -> sum(binomial(n, i) * factorial(i-1), 2:n)
```
Which, using a change of variables `i = n-k+1`, is exactly the current formula:
```julia
sum(k ->  binomial(n, n - k + 1) * factorial(n - k), 1:(n - 1))
```

**Suggested correction**
As correction and enhancement to take into account self-loops, I propose the new following code:
```julia
function ncycles_n_i(n::Int, i::Int, self::Bool = false)
    if !self & (i == 1)
        return 0
    else
        return binomial(big(n), big(i)) * factorial(i - 1)
    end
end

max_simple_cycles(n::Integer, self::Bool = false) = sum(x -> ncycles_n_i(n, x, self), 1:n)
```
The sum starts at 1 for `max_simple_cycles` for the edge case where `n=1` and there are self-loops.

Finally, I must apologize, I introduced the faulty code some time ago...

**Comment by sbromberger** [comment]
Would you mind making a PR for this? I'll do a quick review and merge. 

Thanks!

**Comment by EliasBcd** [comment]
Sure, I will do!

**Comment by sbromberger** [comment]
Closed via #1402. Thank you!
